Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+            /// Indicates that the caller should stop producing elements.
+            case stopProducing
+            /// Indicates that the yielded elements have been dropped because the sequence already terminated.
+            case dropped
+        }
+
+        /// Yields a sequence of new elements to the ``NIOAsyncSequenceProducer``.
+        ///
+        /// If there is an ``NIOAsyncSequenceProducer/AsyncIterator`` awaiting the next element, it will get resumed right away.
+        /// Otherwise, the element will get buffered.
+        ///
+        /// If the ``NIOAsyncSequenceProducer`` is terminated this will drop the elements
+        /// and return ``YieldResult/dropped``.
+        ///
+        /// This can be called more than once and returns to the caller immediately
+        /// without blocking for any awaiting consumption from the iteration.
+        ///
+        /// - Parameter sequence: The sequence to yield.
+        /// - Returns: A ``NIOAsyncSequenceProducer/Source/YieldResult`` that indicates if the yield was successfull
+        /// and if more elements should be produced.
+        @inlinable
+        public func yield<S: Sequence>(_ sequence: S) -> YieldResult where S.Element == Element {
+            self.storage.yield(sequence)
+        }
+
+        /// Yields a new elements to the ``NIOAsyncSequenceProducer``.
+        ///
+        /// If there is an ``NIOAsyncSequenceProducer/AsyncIterator`` awaiting the next element, it will get resumed right away.
+        /// Otherwise, the element will get buffered.
+        ///
+        /// If the ``NIOAsyncSequenceProducer`` is terminated this will drop the elements
+        /// and return ``YieldResult/dropped``.
+        ///
+        /// This can be called more than once and returns to the caller immediately
+        /// without blocking for any awaiting consumption from the iteration.
+        ///
+        /// - Parameter element: The element to yield.
+        /// - Returns: A ``NIOAsyncSequenceProducer/Source/YieldResult`` that indicates if the yield was successfull
+        /// and if more elements should be produced.
+        @inlinable
+        public func yield(_ element: Element) -> YieldResult {
+            self.yield(CollectionOfOne(element))
+        }
+
+        /// Finishes the sequence.
+        ///
+        /// Calling this function signals the sequence that there won't be any subsequent elements yielded.
+        ///
+        /// If there are still buffered elements and there is an ``NIOAsyncSequenceProducer/AsyncIterator`` consuming the sequence,
+        /// then termination of the sequence only happens once all elements have been consumed by the ``NIOAsyncSequenceProducer/AsyncIterator``.
+        /// Otherwise, the buffered elements will be dropped.
+        ///
+        /// - Note: Calling this function more than once has no effect.
+        @inlinable
+        public func finish() {
+            self.storage.finish()
+        }
```

Comment:
Always unhappy about the "finish" wording... :-( but that's what AsyncStream sadly established; LGTM